### PR TITLE
[SCB-1650]discovery schema from instancies directly

### PR DESCRIPTION
**Pull request by liubao68**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean install -Pit` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

File: demo/demo-multi-registries/demo-multi-registries-client/src/main/java/org/apache/servicecomb/demo/registry/IServerEndpoint.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.servicecomb.demo.registry;
+
+public interface IServerEndpoint {
```
Comment (wujimin) [inline demo/demo-multi-registries/demo-multi-registries-client/src/main/java/org/apache/servicecomb/demo/registry/IServerEndpoint.java:20]: no need prefix "I"?

File: foundations/foundation-registry/src/main/java/org/apache/servicecomb/registry/consumer/MicroserviceManager.java [hunk 2]
```diff
@@ -45,11 +47,21 @@ public MicroserviceManager(AppManager appManager, String appId) {
...
-    MicroserviceVersions microserviceVersions = versionsByName.computeIfAbsent(microserviceName, name -> {
-      MicroserviceVersions instance = new MicroserviceVersions(appManager, appId, microserviceName);
-      instance.pullInstances();
-      return instance;
-    });
+    // do not use ConcurrentHashMap computeIfAbsent for versionsByName
+    // because: when create MicroserviceVersions, one creation may depend on another
+    // MicroserviceVersions. And pullInstances will create a new MicroserviceVersions.
+    // Calling ConcurrentHashMap computeIfAbsent inside will get deadlock.
+    MicroserviceVersions microserviceVersions = versionsByName.get(microserviceName);
```
Comment (wujimin) [inline foundations/foundation-registry/src/main/java/org/apache/servicecomb/registry/consumer/MicroserviceManager.java:54]: hashmap is not thread safe, not only for write, but also for read

File: service-registry/registry-local/src/main/java/org/apache/servicecomb/localregistry/LocalRegistryStore.java [hunk 3]
```diff
@@ -83,9 +90,28 @@ public void run() {
     selfMicroserviceInstance.setInstanceId(selfMicroservice.getServiceId());
     selfMicroserviceInstance.setServiceId(selfMicroservice.getServiceId());
 
-    InputStream is = this.getClass().getClassLoader().getResourceAsStream(REGISTRY_FILE_NAME);
-    if (is != null) {
-      initFromData(is);
+    InputStream is = null;
+
+    try {
+      ClassLoader loader = JvmUtils.findClassLoader();
```
Comment (wujimin) [inline service-registry/registry-local/src/main/java/org/apache/servicecomb/localregistry/LocalRegistryStore.java:96]: try org.apache.servicecomb.foundation.common.utils.ResourceUtil#findResourcesBySuffix?

File: service-registry/registry-service-center/src/main/java/org/apache/servicecomb/serviceregistry/registry/AbstractServiceRegistry.java [hunk 4]
```diff
@@ -263,6 +263,7 @@ public void destroy() {
 
   @Override
   // TODO: this is for 3rd party invocation, and a better way can be provided
+  // TODO:  microserviceManager.getVersionsByName() can not be used, will delete this in SCB-1949
```
Comment (wujimin) [inline service-registry/registry-service-center/src/main/java/org/apache/servicecomb/serviceregistry/registry/AbstractServiceRegistry.java:266]: 3rd party is a good mechanism， must not delete it.

File: demo/demo-multi-registries/demo-multi-registries-client/src/main/java/org/apache/servicecomb/demo/registry/IServerEndpoint.java (hunk 1 above)
Comment (liubao68) [reply demo/demo-multi-registries/demo-multi-registries-client/src/main/java/org/apache/servicecomb/demo/registry/IServerEndpoint.java:20]: ServerEndpoint is used for class. Here is a demo, try a quick name like other interfaces in the demo.

File: service-registry/registry-service-center/src/main/java/org/apache/servicecomb/serviceregistry/registry/AbstractServiceRegistry.java (hunk 4 above)
Comment (liubao68) [reply service-registry/registry-service-center/src/main/java/org/apache/servicecomb/serviceregistry/registry/AbstractServiceRegistry.java:266]: OK. I'll fix the problem in next PR without delete the interface.

File: service-registry/registry-local/src/main/java/org/apache/servicecomb/localregistry/LocalRegistryStore.java (hunk 3 above)
Comment (liubao68) [reply service-registry/registry-local/src/main/java/org/apache/servicecomb/localregistry/LocalRegistryStore.java:96]: I see many ohter places code like this. 
'ResourceUtil#findResourcesBySuffix' returns List<URI> and not convinient for open stream.

**Pull request merged**